text/0000-oracles.md

```diff
@@ -0,0 +1,117 @@
+| Status  | Stakeholders | Outcome | Due date |   Owner   |
+| :-----: | :----------: | :-----: | :------: | :-------: |
+| Started |              |         |          | @arjentix |
+
+
+
+## Background
+
+"Oracle" is a common name for different technologies which basically do one thing – bring off-chain date on chain. The possibilities are almost limitless:
+
+1. Real wold currencies
+2. Random numbers generation
+3. Timestamp retrieving
+4. Events in other blockchains
+5. Election results
+6. Geodata i.e. to track delivery chains
+7. Complex calculations
+8. Weather data
+9. Statistical data
+10. The central bank's key rate
+
+Obviosly integrating Oracles into Iroha 2 based blockchains will expand the horizon on possibilites for our users. By providing a clear way to integrate your WASM Trigger with Oracles we encourage people to use Iroha 2 in really complex systems and bring their ideas to life.
+
+## Problem
+
+We need to add Oracles support into Iroha 2. Oracles should be a flexible part of a blockchain and should concret Oracles should not be hardcoded anywhere on our side. There should be a clear mechanism to interract with them and respective documentation.
+
+### Example scenarios
+
+ Scenario 1:
+
+1. End-user interracts with a trigger to buy some asset which is tied to a real world, i.e. gold, for some other asset A;
+2. Trigger calls Oracle 1 to get gold price in dollars, and then calls Oracle 2 to get asset A price in dollars;
+3. Trigger computes the result and transfers money to the buyer.
+
+Scenario 2:
+
+1. Weather sensors have registered a hurricane
+2. Trigger has been activated to transfer insurance benefits to the clients
+
+ Scenario 3:
+
+1. End-user interracts with a trigger, requesting some permission token to be given to their account. This permission can be given only to account which corresponds to University "X" students.
+2. Trigger calls Oracle of University "X" to ensure that account corresponds to a student
+3. On success trigger sends permission token to the account
+
+## Solution
+
+Proposed solution is based on already built parts of Iroha and does not require much efforts to implement.
+It consists of the two parts:
+
+1. `EmitOracleRequest` Iroha Special Instruction;
+2. `OracleRequest` event type.
+
+`EmitOracleRequest`  – is a new ISI that does one thing – emits `OracleRequest` with provided payload.
+
+`OracleRequest`  – special event which directly says that blockchain asks for off-chain data from some Oracle. Such events will be transmitted in event pipeline like all other Iroha 2 events.
```

Comment:
What about encryption of the `OracleRequest` event payload with the oracle's account public key? How much is it possible that the request might contain sensitive data?

Reply:
I think for now we could make payload to be just `Vec<u8>`. So that it can contain any data inlcuding encrypted ones. Actually it can be encrypted not just by oracle account public key, but by anything that oracle requires.

From the other side in public blockchains I don't think there are ways to make sensitive calculations without leaking the information. Because trigger execution details are seen for every peer. Every peer will also see transaction input user has sent to activate this trigger.
So I think really sensitive computation can be done only in private chains where we just trust peer holders